### ISSUE TITLE
Implementing percussion instrument support

### ISSUE THREAD

**Issue by bzier** [cuthbertLab/music21#398]
I encountered issues reading midi files which utilize a drum track (channel 10 - midi percussion). Issue #235 lists percussion as one of the line items for version 6. I am opening this issue to begin the discussion for the solution/implementation design for supporting these types of tracks. @mscuthbert, your insight is much appreciated. Thanks.

**Comment by bzier** [comment]
@mscuthbert ping

**Comment by mscuthbert** [comment]
We've just started work on v6 today

**Comment by jain-sah** [comment]
Can we know an ETA by when the percussion support will be there? We are working on a school project which needs to be finished by summer. So we will wait for this if it gets completed early.

Also, what features exactly are we targetting for v6?

It would be a great help to know what's coming.

**Comment by mscuthbert** [comment]
sorry -- I really can't commit to a timeline on anything; I'll try when I can.  If you absolutely need a feature implemented, there is a paid consulting option: https://web.mit.edu/music21/doc/about/faq.html .

**Comment by nblumoe** [comment]
Would it already be possible to export musicxml or MIDI files from music21, where Musescore would recognise them as drum tracks?

So far I failed to do that. I tried using a `BassDrum` instrument, setting MIDI channel to 10 (and using appropriate pitches, e.g. 35 for the bass drum), adding a percussion clef and even a new class `DrumSet` (the latter allowed me to have Musescore pick up the instrument name at least, but that isn't sufficient).

However, when the musicxml gets imported to Musescore, it doesn't recognize it as a GM Drumset, but will use a Piano sound instead. Once I change the staff properties there to select the `Unpitched Percussion > Drumset` instrument, it works correctly. For the pitched instruments (e.g. Bagpipes, Woodblock) Musescore picks up the instrument correctly

Is there a way to export midi or musicxml from music21 which would be recognised correctly as a drum track, allowing me to play back the track correctly without the manual fix on the instrument in musescore?
I am wondering if there is any attribute I need to specify for this to work or if it cannot be done that easily with `music21` at this stage.

**Comment by nblumoe** [comment]
I figured out that the midiChannel from instruments won't be used in the export, because `channelInstrumentData` from `midi.translate` will just assign new channels from the `acceptableChannelList` ignoring whatever was set on existing instruments.

I was able to hack in support for MIDI channel 10 with a monkey patch. Clearly that hack isn't the right way forward, but just let me know if anyone is interested in it.

**Comment by jacobtylerwalls** [comment]
Thanks for digging in; I think all are in agreement this is the next priority for the project. I've started a draft PR #756 regarding the music21 model for unpitched notes in general, and specifically musicxml import.

I've also done some work with MIDI import of instruments, and there is probably some more work there to preserve all the data (specifically #397 points out some lossiness where we revert to generic `Instrument` objects). I'm willing to pick that up at some point.

I like your approach in this diff, and I have also thought we needed a DrumSet class. What if instead of adding a new `.midiChannel` attribute, we merely check `isinstance(i, instrument.UnpitchedPercussion)` and then write to channel 10 in that case?

If you want to spearhead MIDI export and get a proof of concept pull request working, IMO that would be amazing. How does that sound, @mscuthbert?

**Comment by jacobtylerwalls** [comment]
Ah, this probably needs to be plugged into `autoAssignMidiChannel()` which already has a stub for this, so that we can reuse it in both midi and musicxml export. It still seems to me we should check for isinstance(x, UnpitchedPercussion) rather than trust that the user chose the correct midi channel on `.midiChannel`.

**Comment by nblumoe** [comment]
I think it would still be a good idea to allow for setting MIDI channels on exports explicitly. This is important when a user wants to trigger synths on specific channels. But this could be optional: If the user specified a channel, try to use that. If no explicit channel is defined, use auto assignment (which should use channel 10 for unpitched percussion then).

**Comment by mscuthbert** [comment]
> I think it would still be a good idea to allow for setting MIDI channels on exports explicitly. 

That's a good idea, but it's a separate issue.  Please make one for it, since it's not directly connected to percussion support.

**Comment by jacobtylerwalls** [comment]
@bzier @nblumoe We merged a feature for percussion tracks yesterday that should support the cases you described. Please have a look, and reach out (i.e. open a new issue) if there are parsing failures you can identify on the dev version after #961. Thanks.

(musicXML still needs some work to support unpitched percussion, but the discussion on this thread was focused on MIDI and we have a WIP PR for musicxml in #756 .)
